text/3040-enum-as-to-into.md

```diff
@@ -0,0 +1,270 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+As this RFC proposes using existing, well-known features, little teaching is required. We will introduce the feature to users in two ways: Updates to documentation, and lints which may be upgraded to compile errors.
+
+## Documentation updates
+
+We will update [the enum documentation](https://doc.rust-lang.org/reference/items/enumerations.html#custom-discriminant-values-for-fieldless-enumerations). Initially, it will be updated the text along the lines of:
+
+```diff
+--- a/src/items/enumerations.md
++++ b/src/items/enumerations.md
+@@ -66,8 +66,18 @@ opaque reference to this discriminant can be obtained with the
+ If there is no data attached to *any* of the variants of an enumeration,
+ then the discriminant can be directly chosen and accessed.
+
+-These enumerations can be cast to integer types with the `as` operator by a
+-[numeric cast]. The enumeration can optionally specify which integer each
++If you wish to convert these enumerations to integer types, it is recommended
++that you implement
++[the `Into` trait](https://doc.rust-lang.org/std/convert/trait.Into.html) for
++them. This can be done easily with `#[derive(Into)]`, supplied by the standard
++library.
++
++For legacy reasons, these enumerations can also be cast to integer types with
++the `as` operator by a [numeric cast], but this is not recommended as it can
++have surprising and unsafe results, and may cease to be supported in a
++future version of Rust.
++
++The enumeration can optionally specify which integer each
+ discriminant gets by following the variant name with `=` followed by a [constant
+ expression]. If the first variant in the declaration is unspecified, then it is
+ set to zero. For every other unspecified discriminant, it is set to one higher
+```
+
+In the future, if support is removed for `as` casts, the text may be further updated.
+
+## Lints
+
+Three new lints will be introduced to the language. They will start at `warn` level, and could independently be promoted to `error` level, or full compile errors:
+
+### Likely incorrect casts
+
+For casts where the value being cast is an enum, and the type being cast to is smaller than the enum's repr, a warning will be issued:
+
+```rust
+#[repr(u16)]
+enum Number {
+    Zero,
+    One,
+}
+
+fn main() {
+    let bad = Number::Zero as u8;
+}
+```
+
+```
+warning: unsafe cast
```

Comment:
Nit: please only use "unsafe" to refer to things that can be unsound.  Truncation is by itself always sound, so should use a different word here.

Reply:
I've removed references to the term "safe" from the whole RFC. I'm not sure there's a great direct replacement here, so I varyingly used "predictable", "reliable", "correct", "truncating", and "unexpectedly truncating" depending on context.

Reply:
Thanks!  Much better now.